### PR TITLE
[blob] Try to support resource fork fonts on macOS

### PR DESCRIPTION
**Pull request by khaledhosny**
If the size of opened file is zero, try opening resource fork by appending `/..namedfork/rsrc` to the file name. This is guarded with `__APPLE__` ifdef and uses `_PATH_RSRCFORKSPEC` macro from `sys/paths.h`.

Defining `HB_NO_RSRCFORK` will disable this fallback.

Fixes https://github.com/harfbuzz/harfbuzz/issues/2383

### CODE REVIEW

File: src/hb-blob.cc [hunk 1]
```diff
@@ -520,6 +524,33 @@ _hb_mapped_file_destroy (void *file_)
 }
 #endif
 
+#ifdef _PATH_RSRCFORKSPEC
+static int
+_open_resource_fork(const char *file_name, hb_mapped_file_t *file)
+{
+  size_t name_len = strlen (file_name) + sizeof (_PATH_RSRCFORKSPEC);
+  char *rsrc_name = (char *) malloc (name_len);
+  snprintf (rsrc_name, name_len, "%s%s", file_name, _PATH_RSRCFORKSPEC);
```
Comment (ebraminio) [inline src/hb-blob.cc:533]: maybe easier to write like
snprintf (rsrc_name, name_len, "%s" _PATH_RSRCFORKSPEC, file_name);

File: src/hb-blob.cc [hunk 2]
```diff
@@ -520,6 +524,33 @@ _hb_mapped_file_destroy (void *file_)
...
+
+  int fd = open (rsrc_name, O_RDONLY | O_BINARY, 0);
+  free (rsrc_name);
+
+  if (fd != -1)
+  {
+	struct stat st;
```
Comment (ebraminio) [inline src/hb-blob.cc:540]: 4 spaces indention

**Comment by ebraminio** [comment]
looks much better than the effectively hack I came with! But maybe can be implemented here https://github.com/GNOME/pango/blob/29019094450227a68cacd5170ce7b5735bb2faef/pango/pangofc-fontmap.c#L2958 ? And they won't have two or more faces somehow, right? I mean dfont could have different faces that's why I thought that. Thanks!

File: src/hb-blob.cc (hunk 1 above)
Comment (khaledhosny) [reply src/hb-blob.cc:533]: I was going to write it this way but felt a bit confusing as to which goes where.

File: src/hb-blob.cc [hunk 3]
```diff
@@ -520,6 +524,33 @@ _hb_mapped_file_destroy (void *file_)
 }
 #endif
 
+#ifdef _PATH_RSRCFORKSPEC
+static int
+_open_resource_fork(const char *file_name, hb_mapped_file_t *file)
+{
+  size_t name_len = strlen (file_name) + sizeof (_PATH_RSRCFORKSPEC);
+  char *rsrc_name = (char *) malloc (name_len);
```
Comment (ebraminio) [inline src/hb-blob.cc:532]: an allocation fail check would be nice also

**Comment by khaledhosny** [comment]
> And they won't have two or more faces somehow

The Garamond file that was attached to the issue has three face in it, and accessing them works fine with this patch.

**Comment by jfkthame** [comment]
> looks much better than the effectively hack I came with! But maybe can be implemented here https://github.com/GNOME/pango/blob/29019094450227a68cacd5170ce7b5735bb2faef/pango/pangofc-fontmap.c#L2958 I wonder? And they won't have two or more faces somehow, right? I mean dfont could have different faces that's why I'm suspecting that about this kind of fonts also. Thanks!

A resource-fork font file could contain any number of faces. (OK, there are theoretical limits. But lots, anyhow.)

**Comment by jfkthame** [comment]
> > looks much better than the effectively hack I came with! But maybe can be implemented here https://github.com/GNOME/pango/blob/29019094450227a68cacd5170ce7b5735bb2faef/pango/pangofc-fontmap.c#L2958 I wonder? And they won't have two or more faces somehow, right? I mean dfont could have different faces that's why I'm suspecting that about this kind of fonts also. Thanks!
> 
> A resource-fork font file could contain any number of faces. (OK, there are theoretical limits. But lots, anyhow.)

For example, the Garamond suitcase provided by the reporter contains regular, bold and italic faces.

**Comment by ebraminio** [comment]
> The Garamond file that was attached to the issue has three face in it, and accessing them works fine with this patch.

Oh, didn't see the comment, thanks

File: src/hb-blob.cc (hunk 1 above)
Comment (khaledhosny) [reply src/hb-blob.cc]: snprintf is an overkill indeed, switched to strncpy/strncat.

File: src/hb-blob.cc [hunk 4]
```diff
@@ -520,6 +524,36 @@ _hb_mapped_file_destroy (void *file_)
...
+  if (unlikely (!rsrc_name)) return -1;
+
+  strncpy (rsrc_name, file_name, strlen (file_name));
+  strncat (rsrc_name, _PATH_RSRCFORKSPEC, sizeof (_PATH_RSRCFORKSPEC) - 1);
```
Comment (behdad) [inline src/hb-blob.cc:536]: Since you know how many bytes you copied, you can make this faster by not using `cat`.

File: src/hb-blob.cc [hunk 5]
```diff
@@ -35,6 +35,10 @@
 #include <sys/mman.h>
 #endif /* HAVE_SYS_MMAN_H */
 
+#if !defined(HB_NO_RSRCFORK) && defined(__APPLE__)
```
Comment (behdad) [inline src/hb-blob.cc:38]: `HB_NO_RESOURCE_FORK` please?  No need to be cryptic like that.

**Comment by behdad** [comment]
> am in feeling that we need an API like `hb_open_font_file` or something like that and put such logic there instead injecting the logic in `hb_blob_create_from_file`

Humm.  Yeah that's a bit of an issue that this is going in blob.  But so far I'm happy with it since it doesn't make any assumptions about this being a font file, just that it's a resource fork...

**Comment by ebraminio** [comment]
> Humm. Yeah that's a bit of an issue that this is going in blob. But so far I'm happy with it since it doesn't make any assumptions about this being a font file, just that it's a resource fork...

Understandable. Just for future reference, using the patch,

```
$ src/main Garamond
Opened font file Garamond: 281960 bytes long
DFont Mac Resource Fork
3 font(s) found in file
Font 0 of 3:
  17 table(s) found in font
  Table  0 of 17: LTSH (0x0000f41c+0x000000fd)
  Table  1 of 17: OS/2 (0x0000f51c+0x00000056)
  Table  2 of 17: cmap (0x00015340+0x0000032c)
  Table  3 of 17: cvt  (0x0000f574+0x00000460)
  Table  4 of 17: fpgm (0x0000f9d4+0x0000044b)
  Table  5 of 17: gasp (0x0000fe20+0x00000010)
  Table  6 of 17: glyf (0x0000011c+0x0000f2fe)
  Table  7 of 17: hdmx (0x0000fe30+0x00001e8c)
  Table  8 of 17: head (0x00011cbc+0x00000036)
  Table  9 of 17: hhea (0x00011cf4+0x00000024)
  Table 10 of 17: hmtx (0x00011d18+0x000003e4)
  Table 11 of 17: kern (0x000120fc+0x00000bbe)
  Table 12 of 17: loca (0x00012cbc+0x000003e8)
  Table 13 of 17: maxp (0x000130a4+0x00000020)
  Table 14 of 17: name (0x000130c4+0x00001960)
  Table 15 of 17: post (0x00014a24+0x0000025b)
  Table 16 of 17: prep (0x00014c80+0x000006bd)
Font 1 of 3:
  17 table(s) found in font
  Table  0 of 17: LTSH (0x000103f4+0x000000fd)
  Table  1 of 17: OS/2 (0x000104f4+0x00000056)
  Table  2 of 17: cmap (0x00016778+0x0000032c)
  Table  3 of 17: cvt  (0x0001054c+0x00000446)
  Table  4 of 17: fpgm (0x00010994+0x00000484)
  Table  5 of 17: gasp (0x00010e18+0x00000010)
  Table  6 of 17: glyf (0x0000011c+0x000102d6)
  Table  7 of 17: hdmx (0x00010e28+0x00001e8c)
  Table  8 of 17: head (0x00012cb4+0x00000036)
  Table  9 of 17: hhea (0x00012cec+0x00000024)
  Table 10 of 17: hmtx (0x00012d10+0x000003e4)
  Table 11 of 17: kern (0x000130f4+0x00000bb2)
  Table 12 of 17: loca (0x00013ca8+0x000003e8)
  Table 13 of 17: maxp (0x00014090+0x00000020)
  Table 14 of 17: name (0x000140b0+0x00001d46)
  Table 15 of 17: post (0x00015df8+0x0000025b)
  Table 16 of 17: prep (0x00016054+0x00000721)
Font 2 of 3:
  17 table(s) found in font
  Table  0 of 17: LTSH (0x000105c0+0x000000fd)
  Table  1 of 17: OS/2 (0x000106c0+0x00000056)
  Table  2 of 17: cmap (0x00016b00+0x0000032c)
  Table  3 of 17: cvt  (0x00010718+0x00000632)
  Table  4 of 17: fpgm (0x00010d4c+0x00000481)
  Table  5 of 17: gasp (0x000111d0+0x00000010)
  Table  6 of 17: glyf (0x0000011c+0x000104a2)
  Table  7 of 17: hdmx (0x000111e0+0x00001e8c)
  Table  8 of 17: head (0x0001306c+0x00000036)
  Table  9 of 17: hhea (0x000130a4+0x00000024)
  Table 10 of 17: hmtx (0x000130c8+0x000003e4)
  Table 11 of 17: kern (0x000134ac+0x000008b8)
  Table 12 of 17: loca (0x00013d64+0x000003e8)
  Table 13 of 17: maxp (0x0001414c+0x00000020)
  Table 14 of 17: name (0x0001416c+0x00001d96)
  Table 15 of 17: post (0x00015f04+0x0000025b)
  Table 16 of 17: prep (0x00016160+0x0000099f)
```

```
$ src/main Trebuchet\ MS
Opened font file Trebuchet MS: 542927 bytes long
DFont Mac Resource Fork
4 font(s) found in file
Font 0 of 4:
  20 table(s) found in font
  Table  0 of 20: DSIG (0x0001f7c0+0x00001a9c)
  Table  1 of 20: LTSH (0x00000b20+0x00000244)
  Table  2 of 20: MTfn (0x0001f438+0x00000388)
  Table  3 of 20: OS/2 (0x000001c8+0x00000056)
  Table  4 of 20: VDMX (0x00000d64+0x000005e0)
  Table  5 of 20: cmap (0x00004768+0x00000a1a)
  Table  6 of 20: cvt  (0x00005ba4+0x00000270)
  Table  7 of 20: fpgm (0x00005184+0x0000056d)
  Table  8 of 20: gasp (0x0001f428+0x00000010)
  Table  9 of 20: glyf (0x00006298+0x00017370)
  Table 10 of 20: hdmx (0x00001344+0x00003424)
  Table 11 of 20: head (0x0000014c+0x00000036)
  Table 12 of 20: hhea (0x00000184+0x00000024)
  Table 13 of 20: hmtx (0x00000220+0x00000900)
  Table 14 of 20: kern (0x0001d608+0x000002d6)
  Table 15 of 20: loca (0x00005e14+0x00000482)
  Table 16 of 20: maxp (0x000001a8+0x00000020)
  Table 17 of 20: name (0x0001d8e0+0x00000be7)
  Table 18 of 20: post (0x0001e4c8+0x00000f60)
  Table 19 of 20: prep (0x000056f4+0x000004ad)
Font 1 of 4:
  20 table(s) found in font
  Table  0 of 20: DSIG (0x0001d2f8+0x00001a94)
  Table  1 of 20: LTSH (0x00000b20+0x00000244)
  Table  2 of 20: MTfn (0x0001cf70+0x00000388)
  Table  3 of 20: OS/2 (0x000001c8+0x00000056)
  Table  4 of 20: VDMX (0x00000d64+0x000005e0)
  Table  5 of 20: cmap (0x000049ac+0x00000a1a)
  Table  6 of 20: cvt  (0x00005af8+0x0000015a)
  Table  7 of 20: fpgm (0x000053c8+0x00000561)
  Table  8 of 20: gasp (0x0001cf60+0x00000010)
  Table  9 of 20: glyf (0x000060d8+0x00014bb2)
  Table 10 of 20: hdmx (0x00001344+0x00003668)
  Table 11 of 20: head (0x0000014c+0x00000036)
  Table 12 of 20: hhea (0x00000184+0x00000024)
  Table 13 of 20: hmtx (0x00000220+0x00000900)
  Table 14 of 20: kern (0x0001ac8c+0x000002b2)
  Table 15 of 20: loca (0x00005c54+0x00000482)
  Table 16 of 20: maxp (0x000001a8+0x00000020)
  Table 17 of 20: name (0x0001af40+0x000010bd)
  Table 18 of 20: post (0x0001c000+0x00000f60)
  Table 19 of 20: prep (0x0000592c+0x000001c9)
Font 2 of 4:
  19 table(s) found in font
  Table  0 of 19: DSIG (0x0002100c+0x00001a98)
  Table  1 of 19: LTSH (0x00000b14+0x00000245)
  Table  2 of 19: OS/2 (0x000001b8+0x00000056)
  Table  3 of 19: VDMX (0x00000d5c+0x000005e0)
  Table  4 of 19: cmap (0x000054f8+0x00000a1a)
  Table  5 of 19: cvt  (0x000065fc+0x00000134)
  Table  6 of 19: fpgm (0x00005f14+0x00000561)
  Table  7 of 19: gasp (0x00020ffc+0x00000010)
  Table  8 of 19: glyf (0x00006bb4+0x00018158)
  Table  9 of 19: hdmx (0x0000133c+0x000041bc)
  Table 10 of 19: head (0x0000013c+0x00000036)
  Table 11 of 19: hhea (0x00000174+0x00000024)
  Table 12 of 19: hmtx (0x00000210+0x00000904)
  Table 13 of 19: kern (0x0001ed0c+0x0000025e)
  Table 14 of 19: loca (0x00006730+0x00000484)
  Table 15 of 19: maxp (0x00000198+0x00000020)
  Table 16 of 19: name (0x0001ef6c+0x0000111d)
  Table 17 of 19: post (0x0002008c+0x00000f6f)
  Table 18 of 19: prep (0x00006478+0x00000181)
Font 3 of 4:
  20 table(s) found in font
  Table  0 of 20: DSIG (0x0001f180+0x00001aa4)
  Table  1 of 20: LTSH (0x00000b24+0x00000245)
  Table  2 of 20: MTfn (0x0001edf8+0x00000388)
  Table  3 of 20: OS/2 (0x000001c8+0x00000056)
  Table  4 of 20: VDMX (0x00000d6c+0x000005e0)
  Table  5 of 20: cmap (0x000049b4+0x00000a1a)
  Table  6 of 20: cvt  (0x00005b38+0x00000162)
  Table  7 of 20: fpgm (0x000053d0+0x00000579)
  Table  8 of 20: gasp (0x0001ede8+0x00000010)
  Table  9 of 20: glyf (0x00006120+0x000166e8)
  Table 10 of 20: hdmx (0x0000134c+0x00003668)
  Table 11 of 20: head (0x0000014c+0x00000036)
  Table 12 of 20: hhea (0x00000184+0x00000024)
  Table 13 of 20: hmtx (0x00000220+0x00000904)
  Table 14 of 20: kern (0x0001c808+0x00000258)
  Table 15 of 20: loca (0x00005c9c+0x00000484)
  Table 16 of 20: maxp (0x000001a8+0x00000020)
  Table 17 of 20: name (0x0001ca60+0x00001416)
  Table 18 of 20: post (0x0001de78+0x00000f6f)
  Table 19 of 20: prep (0x0000594c+0x000001ea)
```

**Pull request merged**